File: binding/src/libqexception.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qexception.html
pub const qexception = struct {

    /// New constructs a new QException object.
    pub fn New() QtC.QException {
        return qtc.QException_new();
    }


    pub fn Raise(self: ?*anyopaque, ) void {
        qtc.QException_Raise(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnRaise(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QException_OnRaise(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseRaise(self: ?*anyopaque, ) void {
        qtc.QException_QBaseRaise(@ptrCast(self));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QException_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qunhandledexception.html
pub const qunhandledexception = struct {

    /// New constructs a new QUnhandledException object.
    pub fn New() QtC.QUnhandledException {
        return qtc.QUnhandledException_new();
    }


    pub fn Raise(self: ?*anyopaque, ) void {
        qtc.QUnhandledException_Raise(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnRaise(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QUnhandledException_OnRaise(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseRaise(self: ?*anyopaque, ) void {
        qtc.QUnhandledException_QBaseRaise(@ptrCast(self));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QUnhandledException_Delete(@ptrCast(self));
    }
};

File: binding/src/libqexception.cpp
#include <qexception.h>
#include "libqexception.h"
#include "libqexception.hxx"

QException* QException_new() {
	 return new VirtualQException();
}

QException* QException_new2(const QException* param1) {
	 return new VirtualQException(*param1);
}

void QException_OperatorAssign(QException* self, const QException* param1) {
	self->operator=(*param1);
}

void QException_Raise(const QException* self) {
	self->raise();
}

// Base class handler implementation
void QException_QBaseRaise(const QException* self) {
	auto* vqexception = dynamic_cast<const VirtualQException*>(self);
	if (vqexception && vqexception->isVirtualQException) {
vqexception->setQException_Raise_IsBase(true);
	vqexception->raise();
}
}

// Auxiliary method to allow providing re-implementation
void QException_OnRaise(const QException* self, intptr_t slot) {
	auto* vqexception = dynamic_cast<const VirtualQException*>(self);
	if (vqexception && vqexception->isVirtualQException) {
vqexception->setQException_Raise_Callback(reinterpret_cast<VirtualQException::QException_Raise_Callback>(slot));
}
}

void QException_Delete(QException* self) {
    delete self;
}

QUnhandledException* QUnhandledException_new() {
	 return new VirtualQUnhandledException();
}

QUnhandledException* QUnhandledException_new2(const QUnhandledException* other) {
	 return new VirtualQUnhandledException(*other);
}

QUnhandledException* QUnhandledException_new3(std::exception_ptr exception) {
	 return new VirtualQUnhandledException(exception);
}

void QUnhandledException_Swap(QUnhandledException* self, QUnhandledException* other) {
	self->swap(*other);
}

void QUnhandledException_OperatorAssign(QUnhandledException* self, const QUnhandledException* other) {
	self->operator=(*other);
}

void QUnhandledException_Raise(const QUnhandledException* self) {
	self->raise();
}

std::exception_ptr QUnhandledException_Exception(const QUnhandledException* self) {
	return self->exception();
}

// Base class handler implementation
void QUnhandledException_QBaseRaise(const QUnhandledException* self) {
	auto* vqunhandledexception = dynamic_cast<const VirtualQUnhandledException*>(self);
	if (vqunhandledexception && vqunhandledexception->isVirtualQUnhandledException) {
vqunhandledexception->setQUnhandledException_Raise_IsBase(true);
	vqunhandledexception->raise();
}
}

// Auxiliary method to allow providing re-implementation
void QUnhandledException_OnRaise(const QUnhandledException* self, intptr_t slot) {
	auto* vqunhandledexception = dynamic_cast<const VirtualQUnhandledException*>(self);
	if (vqunhandledexception && vqunhandledexception->isVirtualQUnhandledException) {
vqunhandledexception->setQUnhandledException_Raise_Callback(reinterpret_cast<VirtualQUnhandledException::QUnhandledException_Raise_Callback>(slot));
}
}

void QUnhandledException_Delete(QUnhandledException* self) {
    delete self;
}

